src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,175 @@
+        {
+            if (buildArtifacts.Count() == 0)
+            {
+                return;
+            }
+                foreach (var buildArtifact in buildArtifacts)
+                {
+                    var specificPath = Path.Combine(targetDirectory, buildArtifact.Name);
+                    await DownloadFileContainerAsync(projectId, buildArtifact, specificPath, minimatchFilters, cancellationToken);
+                }
+        }
+
+        private Tuple<long, string> ParseContainerId(string resourceData)
+        {
+            // Example of resourceData: "#/7029766/artifacttool-alpine-x64-Debug"
+            var segments = resourceData.Split('/');
+
+            long containerId;
+            if (segments.Length == 3 && segments[0] == "#" && long.TryParse(segments[1], out containerId))
+            {
+                return new Tuple<long, string>(
+                    containerId,
+                    segments[2]
+                    );
+            }
+            else
+            {
+                var message = $"Resource data value '{resourceData}' was not expected.";
+                throw new ArgumentException(message, "resourceData");
+            }
+        }
+
+        private async Task DownloadFileContainerAsync(Guid projectId, BuildArtifact artifact, string rootPath, IEnumerable<string> minimatchPatterns, CancellationToken cancellationToken)
+        {
+            var containerIdAndRoot = ParseContainerId(artifact.Resource.Data);
+
+            var items = await containerClient.QueryContainerItemsAsync(
+                containerIdAndRoot.Item1,
+                projectId,
+                containerIdAndRoot.Item2
+                );
+
+            IEnumerable<Func<string, bool>> minimatcherFuncs = this.GetMinimatchFuncs(minimatchPatterns);
+            if(minimatcherFuncs !=null && minimatcherFuncs.Count() !=0)
+            {
+                items = this.GetFilteredItems(items, minimatcherFuncs);
+            }
+
+            // Group items by type because we want to create the folder structure first.
+            var groupedItems = from i in items
+                               group i by i.ItemType into g
+                               select g;
+
+            // Now create the folders.
+            var folderItems = groupedItems.SingleOrDefault(i => i.Key == ContainerItemType.Folder);
+            if (folderItems != null)
+            {
+                Parallel.ForEach(folderItems, (folder) =>
+                {
+                    var targetPath = ResolveTargetPath(rootPath, folder);
+                    Directory.CreateDirectory(targetPath);
+                });
+            }
+
+            // Then download the files.
+            var fileItems = groupedItems.SingleOrDefault(i => i.Key == ContainerItemType.File);
+            if (fileItems != null)
+            {
+                Parallel.ForEach(fileItems, (file) =>
```

Comment:
Hi Tanmay, As discussed in the call, you might want to throttle the number of request you make here, since the number of files can be huge in this case like 25k or more. I guess client will handle socket reuse by default, but you might also want to test it out to verify that is indeed the case.

Reply:
Thanks @omeshp . We reduced the parallelism and it worked.